File: problems/101-200/p134.c
#include "lib/utility.h"

long connection(int p1, int p2) {
    int tmp = p1, n = 10;
    while (tmp != 0) {
        n *= 10;
        tmp /= 10;
    }
    long ans = p2;
    while (ans % n != p1) {
        ans += p2;
    }
    return ans;
}

int main(int argc, char *argv[]) {
    start_time();

    int size;
    int *primes = sieve(1000003, &size);

    long sol = 0;
    int pcnt = size / 100;
    for (int i = 2; i < size - 1; i++) {
        if (i % pcnt == 0) {
            fprintf(stderr, "%d%% complete...\n", 100 * i / size);
        }
        sol += connection(primes[i], primes[i + 1]);
    }
    printf("Solution: %li\n", sol);

    end_time();
}

File: problems/101-200/p134.py
from lib.utility import start_time, end_time, mod_inverse
from lib.prime import read_primes
start_time()

primes = read_primes(1000030)[2:]


def connection(i):
    p1, p2 = primes[i], primes[i + 1]
    n = 10 ** len(str(p1))

    # computing a multiplier such that multiplier * p2 == 1 (mod n)
    multiplier = mod_inverse(p2, n)

    # modifying the multiplier so that (p1 * multiplier) * p2 == p1 (mod n)
    multiplier = (p1 * multiplier) % n
    return multiplier * p2


solution = 0
for i in range(len(primes) - 1):
    solution += connection(i)

print('Solution:', solution)
end_time()
